File: 2D array search.py
#many methods
#method-1
#imagine the 2d array as a 1-d array and do binary search
#method-2
#do binary search on columns-when you find bid -check if target lies in that range,if yes do binary search on that row
#method-3
#for each row check if target lies in the range,if yes do binary search on that row

'''O(log(mn)) Time complexity
   O(1) space complexity'''

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        low=0
        high=len(matrix)*len(matrix[0])-1
        while low<=high:
            mid=low+(high-low)//2
            #find row and column
            #why division by this number--
            r=mid//len(matrix[0])
            c=mid%len(matrix[0])
            if matrix[r][c]==target:
                return True
            else:
                if matrix[r][c]>target:
                    high=mid-1
                else:
                    low=mid+1
        return False
        

File: 2darray_search.java
// Time Complexity : 0(log n)
// Space Complexity :
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach
## Problem1 
Search a 2D Matrix(https://leetcode.com/problems/search-a-2d-matrix/)
class Solution {
    public int search(int[] nums, int target) {
        
        if(nums == null || nums.length == 0) return -1;
        
        int low = 0;
        int high = nums.length - 1;
        
        while(low <=high){
            
            int mid = low + (high - low)/2;
            if(nums[mid] == target){
                return mid;
            }
            //left part is sorted
            if(nums[low] <= nums[mid]){
                if(target < nums[mid] && target >= nums[low]){
                    high = mid - 1;
                }
                else{
                    low = mid + 1;
                }
            }
            //right is sorted
            else{
                if(target > nums[mid] && target <= nums[high]){
                    low = mid + 1;
                }else{
                    high = mid -1;
                }
                
            }
        }
        return -1;
    }
}
